### PR TITLE
Remove VSCode references from hat code

### PR DESCRIPTION
**Pull request by pokey**
## Checklist

- [ ] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [ ] I have not broken the cheatsheet

### CODE REVIEW

**Comment by pokey** [comment]
This touches areas of the code that don't have the best test coverage because they're a bit tougher to test. Eg requires triggering settings changes. I did fairly thorough manual testing, but I'd like to drive this for a day or two before we ship it

Once we switch to pure DI we should look into testing this code more thoroughly as it will be easier to mock things

File: src/core/HatAllocator.ts [hunk 1]
```diff
@@ -12,51 +13,31 @@ interface Context {
...
-      ),
-    );
 
     this.disposables.push(
-      vscode.commands.registerCommand(
-        "cursorless.toggleDecorations",
-        this.toggleDecorations,
-      ),
+      ide().hats.onDidChangeAvailableHatStyles(this.addDecorationsDebounced),
+      ide().hats.onDidChangeIsActive(this.addDecorationsDebounced),
```
Comment (pokey) [inline src/core/HatAllocator.ts:25]: `isActive` => `isEnabled`

File: src/ide/vscode/VscodeEvents.ts [hunk 2]
```diff
@@ -43,3 +44,18 @@ function fromVscodeReason(
       return undefined;
   }
 }
+
+export function forwardEvent<S, T>(
```
Comment (pokey) [inline src/ide/vscode/VscodeEvents.ts:48]: doc string

File: src/ide/vscode/hats/VscodeHats.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import { toVscodeRange } from "../../../libs/vscode-common/vscodeUtil";
+import { VscodeHatStyleName } from "../hatStyles.types";
+import VscodeAvailableHatStyles from "../VscodeAvailableHatStyles";
+import VscodeHatDecorationMap from "./VscodeHatDecorationMap";
+import type VscodeIDE from "../VscodeIDE";
+import { VscodeTextEditorImpl } from "../VscodeTextEditorImpl";
+import { HatStyleName } from "../../../libs/common/ide/types/hatStyles.types";
+
+export class VscodeHats implements Hats {
+  private _availableHatStyles: VscodeAvailableHatStyles;
```
Comment (pokey) [inline src/ide/vscode/hats/VscodeHats.ts:19]: rename to eg hat style manager

File: src/ide/vscode/hats/VscodeHats.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        ({ styleName }) => !(styleName in this.availableHatStyles),
+      )
+    ) {
+      // This happens if the user updated multiple settings simultaneously: one
+      // that affects hat rendering and one that alters the set of available
+      // hats. If the render setting notification fires first, then applying the
+      // old hat decorations will fail, so we bail for now, knowing that the
+      // notification for new hat allocations will come immediately after
+      return;
+    }
```
Comment (pokey) [inline src/ide/vscode/hats/VscodeHats.ts:74]: look into testing this kind of thing; don't spend more than 1-2hrs

File: src/libs/common/ide/types/Hats.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+   * @param hatRanges The ranges that all hats should have
+   */
+  setHatRanges(hatRanges: HatRange[]): Promise<void>;
+
+  /**
+   * Mapping from available hat styles to their priorities
+   */
+  readonly availableHatStyles: HatStyleMap;
+  onDidChangeAvailableHatStyles(listener: Listener<[HatStyleMap]>): Disposable;
+
```
Comment (pokey) [inline src/libs/common/ide/types/Hats.ts:37]: just rename to `hatStyles`?  and update docstring to indicate that it should be filtered to the ones enabled by the user

File: src/ide/vscode/hats/VscodeHats.ts (hunk 4 above)
Comment (pokey) [reply src/ide/vscode/hats/VscodeHats.ts]: Decided to bail on this one for now; see #1227

**Review by AndreasArvidsson — APPROVED** [review]

**Pull request merged**